syntax = "proto2";

package apollo.control;

message Vector {
  repeated double columns = 1;
}

message Matrix {
  repeated Vector rows = 1;
}

message Layer {
  enum ActivationFunc {
    RELU = 0;
    TANH = 1;
    SIGMOID = 2;
  }
  optional int32 layer_input_dim = 1;
  optional int32 layer_output_dim = 2;
  optional Matrix layer_input_weight =
      3;  // weight matrix of |input_dim| x |output_dim|
  optional Vector layer_bias = 4;  // vector of bias, size of |output_dim|
  optional ActivationFunc layer_activation_func = 5;
}

message StandardizationFactor {
  optional Vector input_mean = 1;
  optional Vector input_std = 2;
}

message GPModelParam {
  optional int32 input_dim = 1;
  optional int32 output_dim = 2;
  optional int32 input_window_size = 3;
  optional StandardizationFactor standardization_factor = 4;
}

message FnnModel {
  optional int32 dim_input = 1;
  optional Vector input_feature_mean = 2;
  optional Vector input_feature_std = 3;
  optional Vector output_feature_mean = 4;
  optional Vector output_feature_std = 5;
  optional int32 num_layer = 6;
  // num_layer must equal to first layer layer_input_dim
  repeated Layer layer = 7;
  // dim_ouput must equal to last layer layer_output_dim
  optional int32 dim_output = 8;
}
